### PR TITLE
[SPARK-39041][SQL] Mapping Spark Query ResultSet/Schema to TRowSet/TTableSchema directly

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

The PR is mainly refactoring, aiming to support TimestampNTZ/LTZ at the same time in the future for the thrift server.

As we all know, in spark, we have hive dependencies, which can be classified into two types:
- as a client, for accessing hive metastore/storage, etc, which is now v2.3.9, better to stay in a stable low version to be supported by higher hive metastore servers with backward compatibility
- as a server, for being accessed by Hive JDBC/Thrift client, e.g. beeline, which is now v3.1.2, better to have a higher version to support more clients

The problem here is that we now convert spark results to `org.apache.hadoop.hive.serde2.thrift.Type` first and then to `org.apache.hive.service.rpc.thrift.TTypeId`. the former does not have 2 timestamp types, namely, doesn't have `TIMESTAMPLOCALTZ_TYPE`.

To avoid this, we take a shortcut to map spark results to thrift schema and rowset directly. 

Besides, it also can avoid some unnecessary memory copies from type to type.

Most functionalities have been verified in [apache/kyuubi](https://github.com/apache/incubator-kyuubi) for several years.

### Why are the changes needed?

for supporting more spark datatypes through hive jdbc

### How was this patch tested?

existing ut shall be enough

### CODE REVIEW

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 1]
```diff
@@ -373,17 +306,69 @@ private[hive] class SparkExecuteStatementOperation(
...
+  }
+
+  private def toTTypeQualifiers(typ: DataType): TTypeQualifiers = {
+    val ret = new TTypeQualifiers()
+    val qualifiers = typ match {
+      case d: DecimalType =>
+        Map(
+          TCLIServiceConstants.PRECISION -> TTypeQualifierValue.i32Value(d.precision),
+          TCLIServiceConstants.SCALE -> TTypeQualifierValue.i32Value(d.scale)).asJava
+      case _ => Collections.emptyMap[String, TTypeQualifierValue]()
```
Comment (yaooqinn) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:344]: if char/varchar can be seen via resultset metadata, we shall also provide the char length here for client to take advantage of it

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 2]
```diff
@@ -373,17 +306,69 @@ private[hive] class SparkExecuteStatementOperation(
...
+    case ByteType => TTypeId.TINYINT_TYPE
+    case ShortType => TTypeId.SMALLINT_TYPE
+    case IntegerType => TTypeId.INT_TYPE
+    case LongType => TTypeId.BIGINT_TYPE
+    case FloatType => TTypeId.FLOAT_TYPE
+    case DoubleType => TTypeId.DOUBLE_TYPE
+    case StringType => TTypeId.STRING_TYPE
+    case _: DecimalType => TTypeId.DECIMAL_TYPE
+    case DateType => TTypeId.DATE_TYPE
+    // TODO: Shall use TIMESTAMPLOCALTZ_TYPE, keep AS-IS now for
```
Comment (yaooqinn) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:322]: TimestampType keeps being converted to TTypeId.TIMESTAMP_TYPE in this PR

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIServiceClient.java [hunk 3]
```diff
@@ -35,7 +36,7 @@ public SessionHandle openSession(String username, String password)
   }
 
   @Override
-  public RowSet fetchResults(OperationHandle opHandle) throws HiveSQLException {
+  public TRowSet fetchResults(OperationHandle opHandle) throws HiveSQLException {
```
Comment (yaooqinn) [inline sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIServiceClient.java:39]: CLIServiceClient and its child is used in our test only, changes here will not have impacts on real-world clients

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/RowSetUtils.scala [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+      case LongType =>
+        val tI64Value = new TI64Value
+        if (!row.isNullAt(ordinal)) tI64Value.setValue(row.getLong(ordinal))
+        TColumnValue.i64Val(tI64Value)
+
+      case FloatType =>
+        val tDoubleValue = new TDoubleValue
+        if (!row.isNullAt(ordinal)) {
+          val doubleValue = java.lang.Double.valueOf(row.getFloat(ordinal).toString)
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/RowSetUtils.scala:209]: is the `.toString` necessary?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/RowSetUtils.scala [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        ret.add(idx, row.getAs[T](ordinal))
+      }
+      idx += 1
+    }
+    ret
+  }
+
+  private def toTColumnValue(
+      ordinal: Int,
+      row: Row,
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/RowSetUtils.scala:177]: not related to this PR, but it seems more efficient if we can convert `InternalRow` directly to `TRowSet`

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/RowSetUtils.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/RowSetUtils.scala:209]: https://github.com/apache/incubator-kyuubi/issues/1784

precision loss prevention

Comment (cloud-fan) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/RowSetUtils.scala:209]: Can we add a comment to explain it?

Comment (yaooqinn) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/RowSetUtils.scala:217]: updated

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 6]
```diff
@@ -373,17 +306,69 @@ private[hive] class SparkExecuteStatementOperation(
...
+    // unnecessary behavior change
+    case TimestampType => TTypeId.TIMESTAMP_TYPE
+    case TimestampNTZType => TTypeId.TIMESTAMP_TYPE
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:325]: Maybe I'm missing some context here. Looking at `RowSetUtils`, datetime types are use String as result. Where do we use `TTypeId`?

**Review by cloud-fan — APPROVED** [review]
LGTM except for one question

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:325]: this is for rowset schema. jdbc client side can call getString to get the raw string or getObject to get an java object where it is used

**Comment by yaooqinn** [comment]
thanks, merged to master

**Comment by yaooqinn** [comment]
From HIVE-12049
```diff
 else {
+      if (tRowSet.getColumns() != null) { 
+    	  for (TColumn tvalue : tRowSet.getColumns()) {
+    		  columns.add(new ColumnBuffer(tvalue));
+    	  }
+      }
     }

It looks like the client-side code needs to check `tRowSet.getColumns()` not being null when reconstructed the rowset?

**Comment by wangyum** [comment]
Can we fix it through service side. Because it is difficult to update the client version.

**Comment by yaooqinn** [comment]
Hi @wangyum 

Code in this PR 20c9b3dc4fac283f895c8d860b4c6e0144697302 shall work, I guess. But some of them got partially reverted by 91dfc31ff89c5fa024371484f6788234e4a3df10

**Pull request closed without merging**